src/org/stringtemplate/v4/misc/Misc.java

```diff
@@ -157,6 +157,16 @@ public static String replaceEscapedRightAngle(String s) {
                 i += "\\>>".length();
                 continue;
             }
+            if ( c=='\\' ) {
+                String rest = s.substring(i);
+                if (rest.startsWith("\\>")) {
+                    if (rest.length() == 2 || !rest.substring(2).startsWith(">")) {
```

Comment:
This looks like it could have poor performance if `s` is very long. It's better to do these checks without any `substring` operations, because they involve copying. You should try using `charAt(i+1)` and `charAt(i+2)` instead.

Reply:
I was wondering the same, because I wrote this in the same style as the previous escapes in the method.

Googling around, it seems that in Java 6 (which is the oldest supported by ST4), `String#substring` was O(1), but with J7+ it is O(n). I'll make the adjustments for this one. For the rest, it seems you've already addressed it in https://github.com/antlr/stringtemplate4/pull/236/files#diff-4b7cd0c18190068c3a0a57ee6c30a9d4L140-L158 )